Refine device management for kata-agent

ChengyuZhu6 opened the pull request:
Refine device management for kata-agent by:
- move common data constants and functions to kata-types
- introduce a manager for device handlers.

Apokleos commented:
Hi @ChengyuZhu6 is there a related issue or background of this issue ?

ChengyuZhu6 commented:
> Hi @ChengyuZhu6 is there a related issue or background of this issue ?

Hi @Apokleos. Thank you for your suggestion. I will add a related issue. I’m currently refactoring the device code to make device management clearer in kata-agent.

danmihai1 commented:
@ChengyuZhu6 , the changes seem good to me at a quick look. However, because the diff is so large, I fell asleep before finishing my review.

I would prefer multiple smaller diff PRs to be reviewed and tested individually, instead of this large diff. But, if other reviewers are OK with the large PR, that's OK with me too.

ChengyuZhu6 commented:
> the changes seem good to me at a quick look. However, because the diff is so large, I fell asleep before finishing my review.

@danmihai1 Thank you for taking a look at the changes!

Apokleos commented on src/agent/src/device/mod.rs:
```diff
@@ -0,0 +1,1109 @@
+    // log before validation to help with debugging gRPC protocol version differences.
+    info!(
+        logger,
+        "device-vm-path: {}, device-container-path: {}, device-options: {:?}",
```
I think we should keep the two helpful fields: "device-id" and "device-type" when checking the log message.

ChengyuZhu6 replied:
Done.

Apokleos reviewed (commented):
Thx @ChengyuZhu6 Good job! This makes the arch much clearer.
Some comments left ! And I'm most concerned is the relationship between the `memory_device_handler.rs` file and `pmem` and `nvdimm`, or devices' classification issue. Considering their block-oriented usage, IMHO, both pmem and nvdimm might be grouped with block devices, and scsi need do so. The term `"memory_device"` is somewhat generic and could cause confusion with other memory tech(e.g., virtio-mem, balloon etc.).

ChengyuZhu6 commented:
> Some comments left ! And I'm most concerned is the relationship between the `memory_device_handler.rs` file and `pmem` and `nvdimm`, or devices' classification issue. Considering their block-oriented usage, IMHO, both pmem and nvdimm might be grouped with block devices, and scsi need do so. The term `"memory_device"` is somewhat generic and could cause confusion with other memory tech(e.g., virtio-mem, balloon etc.).

Thanks for your comments! Let me address your comments and let you know.

ChengyuZhu6 commented on src/agent/src/device/block_device_handler.rs:
```diff
@@ -0,0 +1,207 @@
+            return Err(anyhow!("Invalid path for virtio mmio blk device"));
+        }
+        if !Path::new(&device.vm_path).exists() {
+            get_virtio_mmio_device_name(ctx.sandbox, &device.vm_path.to_string())
```
Done.

ChengyuZhu6 commented on src/agent/src/device/block_device_handler.rs:
```diff
@@ -0,0 +1,207 @@
+pub struct VirtioBlkCcwDeviceHandler {}
+
+#[derive(Debug)]
+pub struct VirtioMmioDeviceHandler {}
```
Done.

ChengyuZhu6 commented on src/libs/kata-types/src/handler.rs:
```diff
@@ -0,0 +1,49 @@
+    pub fn get_handlers(&self) -> Vec<String> {
+        self.handlers.keys().map(|v| v.to_string()).collect()
+    }
+}
```
Done.

ChengyuZhu6 commented:
> Thx @ChengyuZhu6 Good job! This makes the arch much clearer. Some comments left ! And I'm most concerned is the relationship between the `memory_device_handler.rs` file and `pmem` and `nvdimm`, or devices' classification issue. Considering their block-oriented usage, IMHO, both pmem and nvdimm might be grouped with block devices, and scsi need do so. The term `"memory_device"` is somewhat generic and could cause confusion with other memory tech(e.g., virtio-mem, balloon etc.).

Hey @Apokleos ! I recommend to categorize these devices based on the drivers typically used to manage them in the kata-agent:

| **Device Type**          | **Driver Used**    | **Description**                                                                 |
|--------------------------|--------------------|---------------------------------------------------------------------------------|
| **VirtioBlkPciDevice**   | `virtio-blk`       | Virtio block device over PCI bus.                                               |
| **VirtioBlkCcwDevice**   | `virtio-blk`       | Virtio block device over CCW bus.                                   |
| **VirtioBlkMmioDevice**  | `virtio-blk`       | Virtio block device over MMIO bus.                          |
| **NvdimmDevice**   | `nvdimm`           | Directly passed NVDIMM device for non-volatile memory. |
| **PmemDevice**   | `nvdimm`           | Managed by NVDIMM driver for persistent memory. |
| **ScsiDevice**           | `virtio-scsi`      | SCSI device managed by virtio-scsi controller.                                  |
| **networkDevice**        | Network Drivers    | Depends on the specific network device. |
| **VfioDevice**           |  Vfio Drivers        | VFIO device using vfio drivers for device passthrough.                        |

Please let me know if I make any mistakes.

ChengyuZhu6 commented:
With the PR, I haven’t changed the logic of any device handlers. Instead, I’ve split them into multiple handler files. Additionally, I’ve refactored each commit to combine deletions and additions for each specific device, making it easier for reviewers to review.

ChengyuZhu6 commented on src/libs/kata-types/src/device.rs:
```diff
@@ -0,0 +1,40 @@
+// Copyright (c) 2024 Intel Corporation
+//
+// SPDX-License-Identifier: Apache-2.0
+//
```
@danmihai1 As you can see, other files import the necessary device constants solely from kata-types/devices.rs. The exception is FS_TYPE_HUGETLB, which is only used in `storage/ephemeral_handler.rs`, so I directly move it to `storage/ephemeral_handler.rs`.

ChengyuZhu6 commented on src/agent/src/device/block_device_handler.rs:
```diff
@@ -0,0 +1,207 @@
+    let matcher = VirtioBlkCCWMatcher::new(CCW_ROOT_BUS_PATH, device);
+    let uev = wait_for_uevent(sandbox, matcher).await?;
+    let devname = uev.devname;
+    return match Path::new(SYSTEM_DEV_PATH).join(&devname).to_str() {
```
I see. While this PR is focused on refactoring handlers, if you think it’s necessary and acceptable, I’ll do it.

ChengyuZhu6 commented:
rebased to main and addressed Dan's comments.
